daemon/device-plugin/deviceplugin.go

```diff
@@ -0,0 +1,243 @@
+		return err
+	}
+
+	pluginEndpoint := filepath.Join(pluginapi.DevicePluginPath, sm.socketFile)
+	fmt.Printf("Starting APF Device Plugin server at: %s\n", pluginEndpoint)
+	lis, err := net.Listen("unix", pluginEndpoint)
+	if err != nil {
+		fmt.Printf("Error: Starting APF Device Plugin server failed: %v", err)
+	}
+	sm.grpcServer = grpc.NewServer()
+
+	kubeletEndpoint := filepath.Join("unix:", DeprecatedSockDir, KubeEndPoint)
+
+	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	if err != nil {
+		fmt.Printf("%s device plugin unable connect to Kubelet : %v", resourceName, err)
+		return err
+	}
+	defer conn.Close()
+
+	pluginapi.RegisterDevicePluginServer(sm.grpcServer, sm)
+
+	client := pluginapi.NewRegistrationClient(conn)
+
+	go sm.grpcServer.Serve(lis)
+
+	// Wait for server to start by launching a blocking connection
+	ctx, _ = context.WithTimeout(context.TODO(), 5*time.Second)
+	conn, err = grpc.DialContext(
+		ctx, "unix:"+pluginEndpoint, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+
+	if err != nil {
+		fmt.Printf("error. unable to establish test connection with %s gRPC server: %v", resourceName, err)
+		return err
+	}
+	fmt.Printf("%s device plugin endpoint started serving \n", resourceName)
+	conn.Close()
+
+	ctx = context.Background()
+
+	request := &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     sm.socketFile,
+		ResourceName: resourceName,
+	}
+
+	if _, err = client.Register(ctx, request); err != nil {
+		fmt.Printf("%s device plugin unable to register with Kubelet : %v \n", resourceName, err)
+		return err
+	}
+	fmt.Printf("%s device plugin registered with Kubelet\n", resourceName)
+
+	return nil
+}
+
+func (g *apfManager) ensureConnected() error {
+	if g.client != nil {
+		return nil
+	}
+	dialOptions := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			return net.Dial("unix", addr)
+		}),
+	}
+
+	conn, err := grpc.DialContext(context.Background(), VendorPluginSocketPath, dialOptions...)
+
+	if err != nil {
+		g.log.Error(err, "Failed to connect to vendor plugin")
+		return err
+	}
+	g.conn = conn
+
+	g.client = pb.NewDeviceServiceClient(conn)
+	return nil
+}
+
+func (sm *apfManager) Stop() error {
+	fmt.Printf("Stopping Device Plugin gRPC server..")
+	if sm.grpcServer == nil {
+		return nil
+	}
+
+	sm.grpcServer.Stop()
+	sm.grpcServer = nil
+
+	return sm.cleanup()
+}
+
+func (sm *apfManager) cleanup() error {
+	pluginEndpoint := filepath.Join(pluginapi.DevicePluginPath, sm.socketFile)
+	if err := os.Remove(pluginEndpoint); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	return nil
+}
+
+func (sm *apfManager) PreStartContainer(ctx context.Context, psRqt *pluginapi.PreStartContainerRequest) (*pluginapi.PreStartContainerResponse, error) {
```

Comment:
pretty sure when I implemented DP before, these were required as I was implementing the interface...
so this is why these are here